### ISSUE TITLE
Future features

### ISSUE THREAD

**Issue by mobad** [Adriel-M/Questrade-Rebalancer#1]
Awesome refactor. I was planning on cleaning the script up eventually but you already did!

Glad you got rid of that API, it was a pain to use...

Some features I was thinking would be useful in the future:

- Selling strategies. For example, when you're 5% absolute away from a target ratio or when you're 25% relative off.
- Gradual ETF switching. For example, let's say that a new ETF comes around that is roughly equivalent but has a lower MER. You would only buy the new ETF and only sell the old one and the combined balance would be used to ratio calculations. Ideally you'd be able to define categories like "Canadian equity" and have a list of equivalent ETFs where the first is the highest buy priority and the last is the highest sell priority.
- Tax gain/loss harvesting could likely be part of gradual switching as it would require you to define equivalent ETFs and you could harvest when switching ETFs.

**Comment by Adriel-M** [comment]
Thanks for your feedback!

Regarding point one, I assume the program would be running in the background and waiting for those events to happen? If that's the case I still need to figure out a way to clean up the auth step.

Regarding point two, not sure how that will be beneficial because Questrade has minimum fees. I feel like an all or none selling strategy would be the best (such as the case of tax loss harvesting).

**Comment by mobad** [comment]
Nah, for point one, assuming that you are doing contributions regularly the next time you run it, it would just detect that there is a big deviation and that you should sell.
But some sort of daemon like getpassiv that alerts you when you have extra cash or should sell would also be an interesting idea.

As for auth step improvements... The Questrade API is pretty annoying, it requires you to refresh your key every week or something.
I was thinking of maybe setting up a cronjob to run show accounts every day so it would always keep an active key.
The other option would be to automate the login process and get a new API key or something.

The point of gradual ETF switching is that it would only sell the old stuff when you'd be selling to rebalance anyways.
So you'd be able to skip one fee hit.
But I suppose it would depend on the MER difference. If it was big enough, it would probably be more efficient to sell ASAP.
Though, if you're in a taxable account and haven't been tax gain harvesting and have large gains you could be taking a large capital gain hit as well.

If this is part of tax loss harvesting somehow then when you switch funds then the next time you sell to rebalance it would switch some to the more efficient fund.

I found some somewhat detailed articles and matlab code on tax loss harvesting algorithms you might be interested in.
https://github.com/waimanc/WealthSimple_TaxLossHarvesting
https://research.wealthfront.com/whitepapers/tax-loss-harvesting/

Though I wonder if when doing this for yourself that fees would eat up a lot of the benefits.

**Comment by mobad** [comment]
So now we have that exact situation where an ETF has decreased its MER
https://www.reddit.com/r/PersonalFinanceCanada/comments/7xsi35/vanguard_lowers_management_fee_on_vab/

So VAB should be 0.09% and ZAG is now 0.10%.
So what would be the most efficient way of switching...

A general formula for whether it's worth it is something like.
dollarValueOfETF*(1-merDifference)^years > (sellingFee + bid-ask spread + buyingFee)
But it'll always be worth it depending on how long your timeframe is.

So our options are:
1. Sell all now.
2. Buy new and wait for next sell rebalance and sell what we need to rebalance of the old ETF.
3. Buy new and wait for next sell rebalance and sell all.

Waiting for next sell rebalance will be more efficient if the time to the next rebalance isn't too high but who knows how to predict that.
You could maybe set limits like wait for X months before giving up and selling everything or when the MER fees get too high.

Adding tax loss harvesting in to this complicates things more...
